Source/Core/VideoCommon/CPUCullImpl.h

```diff
@@ -0,0 +1,695 @@
+  output = vfmaq_laneq_f32(output, pos1, vertex, 1);
+  if constexpr (PositionHas3Elems)
+    output = vfmaq_laneq_f32(output, pos2, vertex, 2);
+#else
+  output.x += vertex.x * pos0.x + vertex.y * pos1.x;
+  output.y += vertex.x * pos0.y + vertex.y * pos1.y;
+  output.z += vertex.x * pos0.z + vertex.y * pos1.z;
+  if constexpr (PositionHas3Elems)
+  {
+    output.x += vertex.z * pos2.x;
+    output.y += vertex.z * pos2.y;
+    output.z += vertex.z * pos2.z;
+  }
+#endif
+  output = ApplyMatrix(output, proj0, proj1, proj2, proj3);
+  return output;
+}
+
+template <bool PositionHas3Elems, bool PerVertexPosMtx>
+ATTR_TARGET DOLPHIN_FORCE_INLINE static Vector
+LoadTransformVertex(const u8* data, Vector pos0, Vector pos1, Vector pos2, Vector pos3,
+                    Vector proj0, Vector proj1, Vector proj2, Vector proj3)
+{
+  Vector vertex;
+  if constexpr (PerVertexPosMtx)
+  {
+    // Vertex data layout always starts with posmtx data if available, then position data
+    // Convenient for us, that means offsets are always fixed
+    u32 idx = data[0] & 0x3f;
+    data += sizeof(u32);
+
+    const float* fdata = reinterpret_cast<const float*>(data);
+
+#ifdef USE_NEON
+    LoadTransposedPos(&xfmem.posMatrices[idx * 4], pos0, pos1, pos2, pos3);
+
+    if constexpr (PositionHas3Elems)
+    {
+      vertex = vld1q_f32(fdata);
+    }
+    else
+    {
+      vertex = vcombine_f32(vld1_f32(fdata), vdup_n_f32(0.0f));
+    }
+
+    vertex = TransformVertex<PositionHas3Elems>(vertex, pos0, pos1, pos2, pos3,  //
+                                                proj0, proj1, proj2, proj3);
+#else
+    LoadPos(&xfmem.posMatrices[idx * 4], pos0, pos1, pos2);
+
+    if constexpr (PositionHas3Elems)
+    {
+#if defined(USE_SSE)
+#ifdef USE_SSE41
+      Vector base = _mm_set1_ps(1.0f);
+      vertex = _mm_blend_ps(_mm_loadu_ps(fdata), base, 8);
+#else
+      Vector base = _mm_setr_ps(0.0f, 0.0f, 0.0f, 1.0f);
+      Vector mask = _mm_castsi128_ps(_mm_setr_epi32(-1, -1, -1, 0));
+      vertex = _mm_or_ps(_mm_and_ps(_mm_loadu_ps(fdata), mask), base);
+#endif
+#else
+      vertex.x = fdata[0];
+      vertex.y = fdata[1];
+      vertex.z = fdata[2];
+      vertex.w = 1.0f;
+#endif
+    }
+    else
+    {
+#if defined(USE_SSE)
+      Vector base = _mm_setr_ps(0.0f, 0.0f, 0.0f, 1.0f);
+      vertex = _mm_loadl_pi(base, reinterpret_cast<const __m64*>(fdata));
+#else
+      vertex.x = fdata[0];
+      vertex.y = fdata[1];
+      vertex.z = 0.0f;
+      vertex.w = 1.0f;
+#endif
+    }
+
+    vertex = TransformVertexNoTranspose(vertex, pos0, pos1, pos2, proj0, proj1, proj2, proj3);
+#endif
+  }
+  else
+  {
+    const float* fdata = reinterpret_cast<const float*>(data);
+    if constexpr (PositionHas3Elems)
+    {
+#if defined(USE_SSE)
+      vertex = _mm_loadu_ps(fdata);
+#elif defined(USE_NEON)
+      vertex = vld1q_f32(fdata);
+#else
+      vertex.x = fdata[0];
+      vertex.y = fdata[1];
+      vertex.z = fdata[2];
+#endif
```

Comment:
Doesn't this leave `vertex.w` uninitialized? It looks like it's initialized properly on the `PerVertexPosMtx` path, but not on this one (both here and when `PositionHas3Elems` is false).

Reply:
Yes, unlike `TransformVertexNoTranspose`, `TransformVertex` doesn't actually use those fields.  I could initialize them anyways though.

Reply:
I'd prefer if they were initialized, just to be safe/make things more obvious.